src/core/util/options.js

```diff
@@ -172,7 +172,7 @@ strats.watch = function (parentVal: ?Object, childVal: ?Object): ?Object {
     }
     ret[key] = parent
       ? parent.concat(child)
-      : [child]
+      : Array.isArray(child) ? child : [child]
```

Comment:
IMO the initial fix was correct.The core reason caused bug was because it deal with the child value inconsistently. With the fix it can deal with it correctly.